cmd/otel-allocator/prehook/relabel.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prehook
+
+import (
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/targetscommon"
+)
+
+type RelabelConfigTargetFilter struct {
+	m          sync.RWMutex
+	log        logr.Logger
+	relabelCfg map[string][]*relabel.Config
+}
+
+func NewRelabelConfigTargetFilter(log logr.Logger) Hook {
+	return &RelabelConfigTargetFilter{
+		log:        log,
+		relabelCfg: make(map[string][]*relabel.Config),
+	}
+}
+
+// helper function converts from model.LabelSet to []labels.Label.
+func ConvertLabelToPromLabelSet(lbls model.LabelSet) []labels.Label {
+	newLabels := make([]labels.Label, len(lbls))
+	index := 0
+	for k, v := range lbls {
+		newLabels[index].Name = string(k)
+		newLabels[index].Value = string(v)
+		index++
+	}
+	return newLabels
+}
+
+func (tf *RelabelConfigTargetFilter) Apply(targets map[string]*targetscommon.TargetItem) map[string]*targetscommon.TargetItem {
+	numTargets := len(targets)
+	numRemainingTargets := numTargets
+	// Note: jobNameKey != tItem.JobName (jobNameKey is hashed)
+	for jobNameKey, tItem := range targets {
+		keepTarget := true
+		lset := ConvertLabelToPromLabelSet(tItem.Label)
+		for _, cfg := range tf.relabelCfg[tItem.JobName] {
+			if new_lset := relabel.Process(lset, cfg); new_lset == nil {
+				keepTarget = false
+				break // inner loop
+			} else {
+				lset = new_lset
+			}
+		}
+
+		if !keepTarget {
+			delete(targets, jobNameKey)
+			numRemainingTargets--
+		}
+	}
+
+	tf.log.V(2).Info("Filtering complete", "seen", numTargets, "kept", numRemainingTargets)
+	return targets
+}
+
+func (tf *RelabelConfigTargetFilter) SetConfig(cfgs map[string]*config.Config) {
+	tf.m.Lock()
+	defer tf.m.Unlock()
+	for _, value := range cfgs {
+		for _, scrapeConfig := range value.ScrapeConfigs {
+			tf.relabelCfg[scrapeConfig.JobName] = scrapeConfig.RelabelConfigs
+		}
+	}
+}
```

Comment:
Can you add tests for `SetConfig()`?

Reply:
> Can we transform the cfgs map before the call to SetConfig() and pass the relabelCfg (map[string][]*relabel.Config type) directly?

Hmm the reason I kept it as a config.Config instead was to give more flexibility for the interface. i.e. is there any other valuable info in the config that would be useful to have for a future filtering method?  Would a relabelCfg be unhelpful for certain future filtering methods? I'm really unsure what the best function signature for `SetConfig` should be

> the discovery.Manager translates the map from map[allocatorWatcher.EventSource]*config.Config to map[string]*config.Config but then we aren't using the key values here.

Oh yeah that's super weird I did that conversion to resolve a "import cycle not allowed". importing `allocatorWatcher` would cause a probelm bc `allocatorWatcher` depends on `allocation` which depends on `prehook`, so prehook can't depend on `allocatorWatcher`. This would also be resolved if I just pass the relabelCfg directly as you suggested.

> if a job name disappears after it previously existed, it will still be there in the map (nothing is ever deleted and tf.relabelCfg isn't ever set to an empty map).

Thanks for catching this! Now I am creating a new local map, and assign it to `tf.relabelCfg` then (I think) the stale keys will no longer be in the map after reassignment

> Can you add tests for SetConfig()?

Added a new method to the interface called `GetConfig()` and added `TestSetConfig` which calls GetConfig

Reply:
I think we can change it to be just the relabel.Config type directly as Kristina said. Right now this is what's in that config block https://github.com/prometheus/prometheus/blob/main/config/config.go#L221-L231 i.e. not much we care about.

Reply:
Awesome thank you both for the feedback! Changed the function signature to `SetConfig(map[string][]*relabel.Config)`

The interface is now 

```
type Hook interface {
	Apply(map[string]*targetscommon.TargetItem) map[string]*targetscommon.TargetItem
	SetConfig(map[string][]*relabel.Config)
	GetConfig() map[string][]*relabel.Config
}
```